Fix IRB readline history handling when none is present

rwstauner opened the pull request:
When using irb in singleline mode history navigation keystrokes were resulting in a java error:

```
$ ./mxbuild/truffleruby-jvm/languages/ruby/bin/irb -f --prompt=simple --readline --singleline
>>

truffleruby: an internal exception escaped out of the interpreter,
please report it to https://github.com/oracle/truffleruby/issues.

Index: 605, Size: 605 (java.lang.IndexOutOfBoundsException)
        from java.util.LinkedList.checkElementIndex(LinkedList.java:559)
        from java.util.LinkedList.get(LinkedList.java:480)
        from org.truffleruby.stdlib.readline.MemoryHistory.current(MemoryHistory.java:151)
        from org.graalvm.shadowed.org.jline.reader.impl.LineReaderImpl.moveHistory(LineReaderImpl.java:5369)
        from org.graalvm.shadowed.org.jline.reader.impl.LineReaderImpl.downHistory(LineReaderImpl.java:3117)
        from org.graalvm.shadowed.org.jline.reader.impl.LineReaderImpl.downLineOrHistory(LineReaderImpl.java:3151)
        from org.graalvm.shadowed.org.jline.reader.impl.LineReaderImpl$1.apply(LineReaderImpl.java:3793)
        from org.graalvm.shadowed.org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:665)
        from org.graalvm.shadowed.org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:454)
        from org.truffleruby.stdlib.readline.ReadlineNodes$ReadlineNode.lambda$readline$0(ReadlineNodes.java:153)
        from org.truffleruby.core.thread.ThreadManager.lambda$runUntilResult$4(ThreadManager.java:614)
        from com.oracle.truffle.api.impl.ThreadLocalHandshake$TruffleSafepointImpl.setBlockedBoundary(ThreadLocalHandshake.java:596)
        from com.oracle.truffle.api.impl.ThreadLocalHandshake$TruffleSafepointImpl.setBlockedWithException(ThreadLocalHandshake.java:567)
        from org.truffleruby.core.thread.ThreadManager.runUntilResult(ThreadManager.java:611)
        from org.truffleruby.core.thread.ThreadManager.runUntilResult(ThreadManager.java:590)
        from org.truffleruby.stdlib.readline.ReadlineNodes$ReadlineNode.readline(ReadlineNodes.java:150)
        from org.truffleruby.stdlib.readline.ReadlineNodesFactory$ReadlineNodeFactory$ReadlineNodeGen.executeAndSpecialize(ReadlineNodesFactory.java:524)
        from org.truffleruby.stdlib.readline.ReadlineNodesFactory$ReadlineNodeFactory$ReadlineNodeGen.execute(ReadlineNodesFactory.java:513)
        from org.truffleruby.language.RubyCoreMethodRootNode.execute(RubyCoreMethodRootNode.java:58)
/rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb/input-method.rb:215:in `readline'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb/input-method.rb:215:in `gets'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:539:in `block (2 levels) in eval_input'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:770:in `signal_status'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:538:in `block in eval_input'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb/ruby-lex.rb:284:in `lex'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb/ruby-lex.rb:253:in `block (2 levels) in each_top_level_statement'
        from <internal:core> core/kernel.rb:412:in `loop'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb/ruby-lex.rb:250:in `block in each_top_level_statement'
        from <internal:core> core/throw_catch.rb:36:in `catch'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb/ruby-lex.rb:249:in `each_top_level_statement'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:557:in `eval_input'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:491:in `block in run'
        from <internal:core> core/throw_catch.rb:36:in `catch'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:490:in `run'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/mri/irb.rb:419:in `start'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/lib/gems/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from <internal:core> core/kernel.rb:381:in `load'
        from /rws-data/repos/graal/sdk/mxbuild/linux-amd64/GRAALVM_D83C12B6D6_JAVA17/graalvm-d83c12b6d6-java17-23.1.0-dev/languages/ruby/bin/irb:44:in `<main>'
```

andrykonchin commented:
Thank you for the fix.

I can reproduce the issue locally as well. Good finding.

Semantic of the `current` method isn't documented well but at least out implementation could comply with the way it's implemented in another `History` subclass in JLine:

```java
    public String current() {
        if (index >= size()) {
            return "";
        }
        return items.get(index).line();
    }
```

https://github.com/jline/jline3/blob/ccfb8c54fd9658b85a3d659f3abe85b8b4dcea83/reader/src/main/java/org/jline/reader/impl/history/DefaultHistory.java#L540-L545

And the check `index >= size()` should solve the mentioned above exception I suppose.

I thing that now the method `get` behaves in similar way like implementation in JLine - it raises exception. So I would keep this behaviour (either with explicit checks or with relying on boundary checks in `LinkedList`) and not to return `null`.

```java
    public String get(final int index) {
        int idx = index - offset;
        if (idx >= items.size() || idx < 0) {
            throw new IllegalArgumentException("IndexOutOfBounds: Index:" + idx + ", Size:" + items.size());
        }
        return items.get(idx).line();
    }
```

https://github.com/jline/jline3/blob/ccfb8c54fd9658b85a3d659f3abe85b8b4dcea83/reader/src/main/java/org/jline/reader/impl/history/DefaultHistory.java#L350-L356

rwstauner commented:
Thanks for the pointers!

I moved that check back into `current` since that does address the issue.  I started there, then just saw that `get` was doing the same thing and thought I'd combine them.  I'm not very familiar with JLine, but I'm not surprised that the exception might be expected in some cases so I'm glad to keep the consistency there.

I think in my local testing it was always the forward search that caused the exception so the `>= size()` check should be enough.

rwstauner commented:
I moved the spec to its own file since it was no longer about "arrows"
and I added the `slow` tag like the arrows test had.
Is that how it should be done?
Also, what about the `aarch64` tag?
The similar "arrows" spec had both:
https://github.com/oracle/truffleruby/blob/ced85ef4f41cb8797c74c1722d1f79db455b556b/spec/tags/truffle/irb/arrows_tags.txt#L1-L2

rwstauner commented:
i was able to confirm this on linux amd64 as well as an m1 mac:
```
$ arch
arm64
$ machine
arm64e
```

eregon commented:
I wonder, is there any reason to use `--readline` with `irb`? (instead of the default: reline)
Or maybe `--prompt=simple --singleline` implies `--readline`?
